# Caderno base para análise e pré-processamento do Dataset

# Infopreço dos preços de Combustível no Brasil

In [70]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st
import datetime as dt

In [71]:
infopreco_df = pd.read_csv('infopreco.csv', sep=';')
infopreco_df

CNPJ                                               NOME  \
0       62780000102                                  AUTO POSTO PARATI   
1       62780000102                                  AUTO POSTO PARATI   
2       62780000102                                  AUTO POSTO PARATI   
3       62780000102                                  AUTO POSTO PARATI   
4      300357000195                    POSTO E TRANSPORTADORA PEGORARO   
..              ...                                                ...   
988  93777456000143                                  POSTO PAMPEANO II   
989  93777456000143                                  POSTO PAMPEANO II   
990  98749864000105  FIGUEIRA CENTRO COMERCIO E DERIVADOS DE COMBUS...   
991  98749864000105  FIGUEIRA CENTRO COMERCIO E DERIVADOS DE COMBUS...   
992  98749864000105  FIGUEIRA CENTRO COMERCIO E DERIVADOS DE COMBUS...   

                           ENDEREÇO COMPLEMENTO        BAIRRO  \
0        AVENIDA FILINTO MULLER,645         NaN        CENTRO   
1        AVENIDA FILINTO MULLER,645         NaN        CENTRO   
2        AVENIDA FILINTO MULLER,645         NaN        CENTRO   
3        AVENIDA FILINTO MULLER,645         NaN        CENTRO   
4                RODOVIA BR 163,S/N      KM 786    ZONA RURAL   
..                              ...         ...           ...   
988  AVENIDA PRESIDENTE KENNEDY,185         NaN  VILA BATISTA   
989  AVENIDA PRESIDENTE KENNEDY,185         NaN  VILA BATISTA   
990        RUA BENTO GONCALVES,1295         NaN        CENTRO   
991        RUA BENTO GONCALVES,1295         NaN        CENTRO   
992        RUA BENTO GONCALVES,1295         NaN        CENTRO   

           MUNICÍPIO  UF           PRODUTO VALOR VENDA     DATA CADASTRO  
0        TRES LAGOAS  MS  Gasolina C Comum       4,436  28/06/2018 17:49  
1        TRES LAGOAS  MS            Etanol       3,482  28/06/2018 17:49  
2        TRES LAGOAS  MS       Diesel S500       3,644  28/06/2018 17:49  
3        TRES LAGOAS  MS        Diesel S10       3,734  28/06/2018 17:49  
4              COXIM  MS  Gasolina C Comum       4,349  09/12/2018 19:06  
..               ...  ..               ...         ...               ...  
988  CACAPAVA DO SUL  RS       Diesel S500       3,327  29/06/2018 12:01  
989  CACAPAVA DO SUL  RS        Diesel S10       3,394  29/06/2018 12:01  
990           VIAMAO  RS  Gasolina C Comum       4,399  14/10/2019 16:52  
991           VIAMAO  RS       Diesel S500         3,4  14/10/2019 16:52  
992           VIAMAO  RS        Diesel S10       3,499  14/10/2019 16:52  

[993 rows x 10 columns]

Será que o CNPJ e o nome do posto são realmente importantes para nossa predição acerca do valor do combustível?
Acho que não, então iremos retirá-los do dataset

In [72]:
infopreco_df.drop(['CNPJ', 'NOME'], inplace=True, axis=1)
infopreco_df

ENDEREÇO COMPLEMENTO        BAIRRO  \
0        AVENIDA FILINTO MULLER,645         NaN        CENTRO   
1        AVENIDA FILINTO MULLER,645         NaN        CENTRO   
2        AVENIDA FILINTO MULLER,645         NaN        CENTRO   
3        AVENIDA FILINTO MULLER,645         NaN        CENTRO   
4                RODOVIA BR 163,S/N      KM 786    ZONA RURAL   
..                              ...         ...           ...   
988  AVENIDA PRESIDENTE KENNEDY,185         NaN  VILA BATISTA   
989  AVENIDA PRESIDENTE KENNEDY,185         NaN  VILA BATISTA   
990        RUA BENTO GONCALVES,1295         NaN        CENTRO   
991        RUA BENTO GONCALVES,1295         NaN        CENTRO   
992        RUA BENTO GONCALVES,1295         NaN        CENTRO   

           MUNICÍPIO  UF           PRODUTO VALOR VENDA     DATA CADASTRO  
0        TRES LAGOAS  MS  Gasolina C Comum       4,436  28/06/2018 17:49  
1        TRES LAGOAS  MS            Etanol       3,482  28/06/2018 17:49  
2        TRES LAGOAS  MS       Diesel S500       3,644  28/06/2018 17:49  
3        TRES LAGOAS  MS        Diesel S10       3,734  28/06/2018 17:49  
4              COXIM  MS  Gasolina C Comum       4,349  09/12/2018 19:06  
..               ...  ..               ...         ...               ...  
988  CACAPAVA DO SUL  RS       Diesel S500       3,327  29/06/2018 12:01  
989  CACAPAVA DO SUL  RS        Diesel S10       3,394  29/06/2018 12:01  
990           VIAMAO  RS  Gasolina C Comum       4,399  14/10/2019 16:52  
991           VIAMAO  RS       Diesel S500         3,4  14/10/2019 16:52  
992           VIAMAO  RS        Diesel S10       3,499  14/10/2019 16:52  

[993 rows x 8 columns]

Nossa, parece que a coluna COMPLEMENTO apresenta vários dados faltantes. Vamos então consultar os valores faltantes do nosso dataset

In [73]:
infopreco_df.isnull().sum()

ENDEREÇO           0
COMPLEMENTO      677
BAIRRO            11
MUNICÍPIO          0
UF                 0
PRODUTO            0
VALOR VENDA        0
DATA CADASTRO      0
dtype: int64

E não é que realmente está, vamo então deletar essa coluna também já que ela corresponde a cerca de 2/3 de todos os dados.  Já para a coluna BAIRRO, os dados faltantes podem ser trabalhados

In [74]:
infopreco_df.drop('COMPLEMENTO', inplace=True, axis=1)
infopreco_df

ENDEREÇO        BAIRRO        MUNICÍPIO  UF  \
0        AVENIDA FILINTO MULLER,645        CENTRO      TRES LAGOAS  MS   
1        AVENIDA FILINTO MULLER,645        CENTRO      TRES LAGOAS  MS   
2        AVENIDA FILINTO MULLER,645        CENTRO      TRES LAGOAS  MS   
3        AVENIDA FILINTO MULLER,645        CENTRO      TRES LAGOAS  MS   
4                RODOVIA BR 163,S/N    ZONA RURAL            COXIM  MS   
..                              ...           ...              ...  ..   
988  AVENIDA PRESIDENTE KENNEDY,185  VILA BATISTA  CACAPAVA DO SUL  RS   
989  AVENIDA PRESIDENTE KENNEDY,185  VILA BATISTA  CACAPAVA DO SUL  RS   
990        RUA BENTO GONCALVES,1295        CENTRO           VIAMAO  RS   
991        RUA BENTO GONCALVES,1295        CENTRO           VIAMAO  RS   
992        RUA BENTO GONCALVES,1295        CENTRO           VIAMAO  RS   

              PRODUTO VALOR VENDA     DATA CADASTRO  
0    Gasolina C Comum       4,436  28/06/2018 17:49  
1              Etanol       3,482  28/06/2018 17:49  
2         Diesel S500       3,644  28/06/2018 17:49  
3          Diesel S10       3,734  28/06/2018 17:49  
4    Gasolina C Comum       4,349  09/12/2018 19:06  
..                ...         ...               ...  
988       Diesel S500       3,327  29/06/2018 12:01  
989        Diesel S10       3,394  29/06/2018 12:01  
990  Gasolina C Comum       4,399  14/10/2019 16:52  
991       Diesel S500         3,4  14/10/2019 16:52  
992        Diesel S10       3,499  14/10/2019 16:52  

[993 rows x 7 columns]

Vamos ver quais são esses bairros faltantes, se tivermos sorte eles só possuem algum erro de digitação

In [75]:
infopreco_df[infopreco_df.isna().any(axis=1)]

ENDEREÇO BAIRRO           MUNICÍPIO  UF  \
37   RODOVIA BR 040 KM 102,S N    NaN          CRISTALINA  GO   
38   RODOVIA BR 040 KM 102,S N    NaN          CRISTALINA  GO   
39   RODOVIA BR 040 KM 102,S N    NaN          CRISTALINA  GO   
40   RODOVIA BR 040 KM 102,S N    NaN          CRISTALINA  GO   
307      FAZENDA GATURIANO,S N    NaN  DOM EXPEDITO LOPES  PI   
308      FAZENDA GATURIANO,S N    NaN  DOM EXPEDITO LOPES  PI   
309      FAZENDA GATURIANO,S N    NaN  DOM EXPEDITO LOPES  PI   
771          RODOVIA MG 5,KM 6    NaN              SABARA  MG   
772          RODOVIA MG 5,KM 6    NaN              SABARA  MG   
773          RODOVIA MG 5,KM 6    NaN              SABARA  MG   
774          RODOVIA MG 5,KM 6    NaN              SABARA  MG   

              PRODUTO VALOR VENDA     DATA CADASTRO  
37   Gasolina C Comum        4,76  07/02/2020 17:26  
38             Etanol        3,39  07/02/2020 17:26  
39        Diesel S500        3,69  07/02/2020 17:26  
40         Diesel S10        3,73  07/02/2020 17:27  
307  Gasolina C Comum       4,579  29/08/2019 10:02  
308       Diesel S500       3,499  29/08/2019 10:02  
309        Diesel S10       3,539  29/08/2019 10:02  
771       Diesel S500       3,224  11/03/2019 15:26  
772  Gasolina C Comum       4,157  11/03/2019 15:26  
773            Etanol        2,88  11/03/2019 15:27  
774        Diesel S10       3,289  11/03/2019 15:26

Olhe só, os dados com erro são subsequentes, ou seja, provavelmente resultantes de erro humano. Podemos facilmente resolver este problema classificando como RODOVIA e ZONA RURAL

In [76]:
infopreco_df['BAIRRO'].iloc[37:41] = 'RODOVIA'
infopreco_df['BAIRRO'].iloc[771:775] = 'RODOVIA'
infopreco_df['BAIRRO'].iloc[307:310] = 'ZONA RURAL'

Conferindo se todos os dados faltantes foram resolvidos

In [77]:
infopreco_df.isnull().sum()

ENDEREÇO         0
BAIRRO           0
MUNICÍPIO        0
UF               0
PRODUTO          0
VALOR VENDA      0
DATA CADASTRO    0
dtype: int64

Vamos conferir os tipos do dataset

In [78]:
infopreco_df.dtypes

ENDEREÇO         object
BAIRRO           object
MUNICÍPIO        object
UF               object
PRODUTO          object
VALOR VENDA      object
DATA CADASTRO    object
dtype: object

Que estranho o VALOR VENDA ser um objeto, isso deve ocorrer porque os valores estão formatados no padrão brasileiro, com ','. Vamos então arrumar essa formatação e transformá-lo em float

In [79]:
infopreco_df['VALOR VENDA'] = infopreco_df['VALOR VENDA'].apply(lambda x: str(x).replace(',','.'))
infopreco_df['VALOR VENDA'] = infopreco_df['VALOR VENDA'].astype('float64')
infopreco_df

ENDEREÇO        BAIRRO        MUNICÍPIO  UF  \
0        AVENIDA FILINTO MULLER,645        CENTRO      TRES LAGOAS  MS   
1        AVENIDA FILINTO MULLER,645        CENTRO      TRES LAGOAS  MS   
2        AVENIDA FILINTO MULLER,645        CENTRO      TRES LAGOAS  MS   
3        AVENIDA FILINTO MULLER,645        CENTRO      TRES LAGOAS  MS   
4                RODOVIA BR 163,S/N    ZONA RURAL            COXIM  MS   
..                              ...           ...              ...  ..   
988  AVENIDA PRESIDENTE KENNEDY,185  VILA BATISTA  CACAPAVA DO SUL  RS   
989  AVENIDA PRESIDENTE KENNEDY,185  VILA BATISTA  CACAPAVA DO SUL  RS   
990        RUA BENTO GONCALVES,1295        CENTRO           VIAMAO  RS   
991        RUA BENTO GONCALVES,1295        CENTRO           VIAMAO  RS   
992        RUA BENTO GONCALVES,1295        CENTRO           VIAMAO  RS   

              PRODUTO  VALOR VENDA     DATA CADASTRO  
0    Gasolina C Comum        4.436  28/06/2018 17:49  
1              Etanol        3.482  28/06/2018 17:49  
2         Diesel S500        3.644  28/06/2018 17:49  
3          Diesel S10        3.734  28/06/2018 17:49  
4    Gasolina C Comum        4.349  09/12/2018 19:06  
..                ...          ...               ...  
988       Diesel S500        3.327  29/06/2018 12:01  
989        Diesel S10        3.394  29/06/2018 12:01  
990  Gasolina C Comum        4.399  14/10/2019 16:52  
991       Diesel S500        3.400  14/10/2019 16:52  
992        Diesel S10        3.499  14/10/2019 16:52  

[993 rows x 7 columns]

In [80]:
infopreco_df.describe()

VALOR VENDA
count   993.000000
mean     10.622513
std      53.436981
min       0.001000
25%       3.499000
50%       3.799000
75%       4.299000
max     550.000000

Como assim temos valores máximos de 550 reais? Alguém errou o lugar para colocar a virgula ai. Vamos resolver isso

In [81]:
for c in infopreco_df['VALOR VENDA'].unique():
    if c >= 10:
        print('Valor antigo',c)

Valor antigo 499.0
Valor antigo 389.0
Valor antigo 349.0
Valor antigo 359.0
Valor antigo 550.0
Valor antigo 360.0
Valor antigo 445.0
Valor antigo 455.0
Valor antigo 427.0
Valor antigo 377.0
Valor antigo 505.0
Valor antigo 405.0
Valor antigo 409.0
Valor antigo 400.0


In [82]:
infopreco_df.loc[infopreco_df['VALOR VENDA'] > 10, 'VALOR VENDA'] = infopreco_df['VALOR VENDA']/100

In [83]:
infopreco_df.describe()

VALOR VENDA
count   993.000000
mean      3.865987
std       0.651861
min       0.001000
25%       3.499000
50%       3.799000
75%       4.298000
max       5.500000

Endereço é algo muito amplo não? E se conseguissimos apenas extrair se é uma Rua, rodovia ou avenida? E depois deletar o ENDEREÇO

In [84]:
infopreco_df['ACESSO'] = ""
infopreco_df.loc[infopreco_df['ACESSO']=='', 'ACESSO'] = infopreco_df['ENDEREÇO'].str.split().str.get(0)

In [86]:
infopreco_df.drop('ENDEREÇO', inplace=True, axis=1)
infopreco_df

BAIRRO        MUNICÍPIO  UF           PRODUTO  VALOR VENDA  \
0          CENTRO      TRES LAGOAS  MS  Gasolina C Comum        4.436   
1          CENTRO      TRES LAGOAS  MS            Etanol        3.482   
2          CENTRO      TRES LAGOAS  MS       Diesel S500        3.644   
3          CENTRO      TRES LAGOAS  MS        Diesel S10        3.734   
4      ZONA RURAL            COXIM  MS  Gasolina C Comum        4.349   
..            ...              ...  ..               ...          ...   
988  VILA BATISTA  CACAPAVA DO SUL  RS       Diesel S500        3.327   
989  VILA BATISTA  CACAPAVA DO SUL  RS        Diesel S10        3.394   
990        CENTRO           VIAMAO  RS  Gasolina C Comum        4.399   
991        CENTRO           VIAMAO  RS       Diesel S500        3.400   
992        CENTRO           VIAMAO  RS        Diesel S10        3.499   

        DATA CADASTRO   ACESSO  
0    28/06/2018 17:49  AVENIDA  
1    28/06/2018 17:49  AVENIDA  
2    28/06/2018 17:49  AVENIDA  
3    28/06/2018 17:49  AVENIDA  
4    09/12/2018 19:06  RODOVIA  
..                ...      ...  
988  29/06/2018 12:01  AVENIDA  
989  29/06/2018 12:01  AVENIDA  
990  14/10/2019 16:52      RUA  
991  14/10/2019 16:52      RUA  
992  14/10/2019 16:52      RUA  

[993 rows x 7 columns]

In [87]:
infopreco_df['ACESSO'].value_counts()

AVENIDA     449
RUA         254
RODOVIA     204
ESTRADA      34
PRACA         7
ACESSO        6
FAZENDA       6
SETOR         5
ETR           5
VIA           4
LINHA         4
LOTE          4
SERVIDAO      3
ALAMEDA       3
QUADRA        3
DISTRITO      2
Name: ACESSO, dtype: int64